### ISSUE TITLE
session opening reminders timeout

### ISSUE THREAD

**Issue by damithc** [TEAMMATES/teammates#6018]
Getting a few of these. This is more critical than closing reminders. Possibly due to large courses?
```
08:02:01.058
teammates.logic.automated.AutomatedRemindersServlet logMessage: TEAMMATESLOG|||feedbackSessionOpeningReminders|||send opening reminders|||true|||Auto|||Unknown|||Unknown|||Unknown|||Generating reminders for opening feedback sessions.|||/feedbackSessionOpeningReminders|||Unknown%2016082108020158
08:02:23.150
com.google.appengine.api.datastore.BaseQueryResultsSource logChunkSizeWarning: This query does not have a chunk size set in FetchOptions and has returned over 1000 results.  If result sets of this size are common for this query, consider setting a chunk size to improve performance.
  To disable this warning set the following system property in appengine-web.xml (the value of the property doesn't matter): 'appengine.datastore.disableChunkSizeWarning'
08:03:12.046
/feedbackSessionOpeningReminders
org.datanucleus.store.appengine.query.QueryExceptionWrappers$JDOQueryTimeoutException
	at org.datanucleus.store.appengine.query.QueryExceptionWrappers$2.get(QueryExceptionWrappers.java:88)
	at org.datanucleus.store.appengine.query.RuntimeExceptionWrappingIterator.hasNext(RuntimeExceptionWrappingIterator.java:105)
	at org.datanucleus.store.appengine.query.LazyResult$LazyAbstractListIterator.computeNext(LazyResult.java:210)
	at org.datanucleus.store.appengine.query.AbstractIterator.tryToComputeNext(AbstractIterator.java:132)
	at org.datanucleus.store.appengine.query.AbstractIterator.hasNext(AbstractIterator.java:127)
	at org.datanucleus.store.appengine.query.LazyResult$AbstractListIterator.hasNext(LazyResult.java:169)
	at teammates.storage.api.FeedbackSessionsDb.getFeedbackSessionsWithUnsentOpenEmail(FeedbackSessionsDb.java:182)
	at teammates.logic.core.FeedbackSessionsLogic.getFeedbackSessionsWhichNeedOpenEmailsToBeSent(FeedbackSessionsLogic.java:1117)
	at teammates.logic.core.FeedbackSessionsLogic.scheduleFeedbackSessionOpeningEmails(FeedbackSessionsLogic.java:1614)
	at teammates.logic.automated.FeedbackSessionOpeningRemindersServlet.doGet(FeedbackSessionOpeningRemindersServlet.java:21)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:617)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at teammates.storage.datastore.DatastoreFilter.doFilter(DatastoreFilter.java:29)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.runtime.jetty.SaveSessionFilter.doFilter(SaveSessionFilter.java:37)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:50)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at com.google.apphosting.runtime.jetty.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:260)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
	at com.google.apphosting.runtime.jetty.RpcRequestParser.parseAvailable(RpcRequestParser.java:78)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:148)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:504)
	at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:446)
	at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:453)
	at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:276)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:312)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:304)
	at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:450)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:235)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.datanucleus.store.appengine.DatastoreExceptionTranslator$1: The datastore operation timed out, or the data was temporarily unavailable.
	at org.datanucleus.store.appengine.DatastoreExceptionTranslator.wrapDatastoreTimeoutExceptionForQuery(DatastoreExceptionTranslator.java:65)
	at org.datanucleus.store.appengine.query.QueryExceptionWrappers$1.get(QueryExceptionWrappers.java:55)
	at org.datanucleus.store.appengine.query.QueryExceptionWrappers$2.get(QueryExceptionWrappers.java:86)
	... 45 more
Caused by: com.google.appengine.api.datastore.DatastoreTimeoutException: The datastore operation timed out, or the data was temporarily unavailable.
	at com.google.appengine.api.datastore.DatastoreApiHelper.translateError(DatastoreApiHelper.java:64)
	at com.google.appengine.api.datastore.DatastoreApiHelper$1.convertException(DatastoreApiHelper.java:129)
	at com.google.appengine.api.utils.FutureWrapper.get(FutureWrapper.java:97)
	at com.google.appengine.api.datastore.FutureHelper.getInternal(FutureHelper.java:76)
	at com.google.appengine.api.datastore.FutureHelper.quietGet(FutureHelper.java:36)
	at com.google.appengine.api.datastore.BaseQueryResultsSource.loadMoreEntities(BaseQueryResultsSource.java:197)
	at com.google.appengine.api.datastore.BaseQueryResultsSource.loadMoreEntities(BaseQueryResultsSource.java:171)
	at com.google.appengine.api.datastore.QueryResultIteratorImpl.ensureLoaded(QueryResultIteratorImpl.java:148)
	at com.google.appengine.api.datastore.QueryResultIteratorImpl.hasNext(QueryResultIteratorImpl.java:64)
	at com.google.appengine.api.datastore.PreparedMultiQuery$FilteredMultiQueryIterator.computeNext(PreparedMultiQuery.java:249)
	at com.google.appengine.api.datastore.PreparedMultiQuery$FilteredMultiQueryIterator.computeNext(PreparedMultiQuery.java:185)
	at com.google.appengine.api.datastore.AbstractIterator.tryToComputeNext(AbstractIterator.java:46)
	at com.google.appengine.api.datastore.AbstractIterator.hasNext(AbstractIterator.java:41)
	at com.google.appengine.api.datastore.QueryResultIteratorDelegator.hasNext(QueryResultIteratorDelegator.java:38)
	at org.datanucleus.store.appengine.query.RuntimeExceptionWrappingIterator$1.get(RuntimeExceptionWrappingIterator.java:50)
	at org.datanucleus.store.appengine.query.RuntimeExceptionWrappingIterator$1.get(RuntimeExceptionWrappingIterator.java:46)
	at org.datanucleus.store.appengine.query.QueryExceptionWrappers$1.get(QueryExceptionWrappers.java:51)
	... 46 more

**Comment by damithc** [comment]
Needs an immediate fix. This is worse than closing reminders not working.
@wkurniawan07

**Comment by damithc** [comment]
Perhaps add the whole job into the task queue which will then spawn more tasks for each session, and so on? From what I remember, tasks get a 10 minute window to execute, unlike normal requests that get only 1 minute.

**Comment by damithc** [comment]
Looks like we have lot of legacy sessions with 'sentOpenEmail' set to false for some reason.
We can either add an additional filter to restrict results to sessions with open time in next 48 hours and/or correct the sessions with the wrong flag.

**Comment by damithc** [comment]
![image](https://cloud.githubusercontent.com/assets/1673303/17836445/baf6465c-67c5-11e6-879d-0f24ae774258.png)

**Comment by damithc** [comment]
Confirmed filtering by endDate solves the problem. I deployed a hack to test it out. 
Here's the hack:
```java
    @SuppressWarnings("unchecked")
    private List<FeedbackSession> getFeedbackSessionEntitiesWithUnsentOpenEmail() {
        Query q = getPm().newQuery(FeedbackSession.class);
        q.declareParameters("java.util.Date yesterday, boolean sentParam");
        q.setFilter("endTime > yesterday && sentOpenEmail == sentParam");
        
        SimpleDateFormat sdf = new SimpleDateFormat("dd/MM/yyyy");
        Date d = null;
        try {
            d = sdf.parse("01/08/2016");
        } catch (ParseException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
        
        return (List<FeedbackSession>) q.execute(d, false);
    }
```
Private sessions are filtered out at this method (cannot have more than one inequality in the query)
```java
public List<FeedbackSessionAttributes> getFeedbackSessionsWhichNeedOpenEmailsToBeSent() {
        List<FeedbackSessionAttributes> sessions =
                fsDb.getFeedbackSessionsWithUnsentOpenEmail();
        
        List<FeedbackSessionAttributes> sessionsToSendEmailsFor =
                new ArrayList<FeedbackSessionAttributes>();

        for (FeedbackSessionAttributes session : sessions) {
            if (session.getFeedbackSessionType() != FeedbackSessionType.PRIVATE && session.isOpened()) {
                sessionsToSendEmailsFor.add(session);
            }
        }
        
        log.info("Number of sessions to send opening emails " + sessionsToSendEmailsFor.size());
        
        return sessionsToSendEmailsFor;
    }
```

**Comment by wkurniawan07** [comment]
I'll try to apply that in all four (or more?) reminder emails.

**Comment by damithc** [comment]
You may want to use a 96 hour window for filtering (i.e. `t-48` .. `t+48`) so that any missed reminders in the past 48 will be taken care of as well.

**Comment by damithc** [comment]
Is this going to be fixed soon? Currently I have to run my hack manually every hour.

**Comment by wkurniawan07** [comment]
Fixed in #6029
